src/mlj_flux.jl

```diff
@@ -0,0 +1,344 @@
+        finaliser,
+        optimiser,
+        loss,
+        epochs,
+        batch_size,
+        lambda,
+        alpha,
+        rng,
+        optimiser_changes_trigger_retraining,
+        acceleration,
+        likelihood,
+        subset_of_weights,
+        subnetwork_indices,
+        hessian_structure,
+        backend,
+        σ,
+        μ₀,
+        P₀,
+        link_approx,
+        fit_params,
+        nothing,
+    )
+end
+
+function MLJFlux.shape(model::LaplaceApproximation, X, y)
+    X = X isa Matrix ? Tables.table(X) : X
+    levels = MMI.classes(y[1])
+    n_output = length(levels)
+    n_input = length(Tables.schema(X).names)
+    return (n_input, n_output)
+end
+
+function MLJFlux.build(model::LaplaceApproximation, rng, shape)
+    # Construct the chain
+    chain = Flux.Chain(MLJFlux.build(model.builder, rng, shape...), model.finaliser)
+    # Construct Laplace model and store it in the model object
+    model.la = Laplace(
+        chain;
+        likelihood=model.likelihood,
+        subset_of_weights=model.subset_of_weights,
+        subnetwork_indices=model.subnetwork_indices,
+        hessian_structure=model.hessian_structure,
+        backend=model.backend,
+        σ=model.σ,
+        μ₀=model.μ₀,
+        P₀=model.P₀,
+    )
+    return chain
+end
+
+function MLJFlux.fitresult(model::LaplaceApproximation, chain, y)
+    return (chain, model.la, MMI.classes(y[1]))
+end
+
+function MLJFlux.train!(model::LaplaceApproximation, penalty, chain, optimiser, X, y)
+    loss = model.loss
+    n_batches = length(y)
+    training_loss = zero(Float32)
+    for i in 1:n_batches
+        parameters = Flux.params(chain)
+        gs = Flux.gradient(parameters) do
+            yhat = chain(X[i])
+            batch_loss = loss(yhat, y[i]) + penalty(parameters) / n_batches
+            training_loss += batch_loss
+            return batch_loss
+        end
+        Flux.update!(optimiser, parameters, gs)
+    end
+    return training_loss / n_batches
+end
+
+function MLJFlux.fit!(
+    model::LaplaceApproximation, penalty, chain, optimiser, epochs, verbosity, X, y
+)
+    loss = model.loss
+
+    # intitialize and start progress meter:
+    meter = Progress(
+        epochs + 1;
+        dt=0,
+        desc="Optimising neural net:",
+        barglyphs=BarGlyphs("[=> ]"),
+        barlen=25,
+        color=:yellow,
+    )
+    verbosity != 1 || next!(meter)
+
+    # initiate history:
+    n_batches = length(y)
+
+    parameters = Flux.params(chain)
+
+    # initial loss:
+    losses = (
+        loss(chain(X[i]), y[i]) + penalty(parameters) / n_batches for i in 1:n_batches
+    )
+    history = [mean(losses)]
+
+    for i in 1:epochs
+        current_loss = MLJFlux.train!(
+            model::MLJFlux.MLJFluxModel, penalty, chain, optimiser, X, y
+        )
+        verbosity < 2 || @info "Loss is $(round(current_loss; sigdigits=4))"
+        verbosity != 1 || next!(meter)
+        push!(history, current_loss)
+    end
+
```

Comment:
So as I understand, correct me if I am wrong, what is done here is: `MLJFlux.fit!` fits the actual NN model (instantiated by the builder) in this for-loop, and then we fit the LA.

Reply:
Quite right you are, this function fits the nn with the specified epoch number, the process which we would otherwise have outside the function.